oscore-cf/src/main/java/org/eclipse/californium/oscore/OSSerializer.java

```diff
@@ -0,0 +1,382 @@
+						aad.Add(version);
+						aad.Add(algorithms);
+						aad.Add(ctx.getRecipientId());
+						aad.Add(processPartialIV(seq));
+						return aad.EncodeToBytes();
+					} else {
+						LOGGER.error(Error.OPTIONSET_NULL);
+						throw new NullPointerException(Error.OPTIONSET_NULL);
+					}
+				} else {
+					LOGGER.error(Error.CTX_NULL);
+					throw new NullPointerException(Error.CTX_NULL);
+				}
+			} else {
+				LOGGER.error(Error.SEQ_NBR_INVALID);
+				throw new IllegalArgumentException(Error.SEQ_NBR_INVALID);
+			}
+		} else {
+			LOGGER.error(Error.WRONG_VERSION_NBR);
+			throw new IllegalArgumentException(Error.WRONG_VERSION_NBR);
+		}
+	}
+
+	/**
+	 * Generates the nonce.
+	 * 
+	 * @param partialIV
+	 * @param senderID
+	 * @param commonIV
+	 * @param nonceLength the algorithm dependent length of nonce
+	 * @return the generated nonce or null if either one of the input parameters
+	 *         are null
+	 * @throws OSException if any of the parameters are unvalid
+	 */
+	public static byte[] nonceGeneration(byte[] partialIV, byte[] senderID, byte[] commonIV, int nonceLength)
+			throws OSException {
+		if (partialIV != null) {
+			if (senderID != null) {
+				if (commonIV != null) {
+					if (nonceLength > 0) {
+						int s = senderID.length;
+						int zeroes = 5 - partialIV.length;
+
+						if (zeroes > 0) {
+							partialIV = leftPaddingZeroes(partialIV, zeroes);
+						}
+
+						zeroes = (nonceLength - 6) - senderID.length;
+
+						if (zeroes > 0) {
+							senderID = leftPaddingZeroes(senderID, zeroes);
+						}
+
+						zeroes = nonceLength - commonIV.length;
+
+						if (zeroes > 0) {
+							commonIV = leftPaddingZeroes(commonIV, zeroes);
+						}
+
+						byte[] tmp = new byte[1 + senderID.length + partialIV.length];
+						tmp[0] = (byte) s;
+						System.arraycopy(senderID, 0, tmp, 1, senderID.length);
+						System.arraycopy(partialIV, 0, tmp, senderID.length + 1, partialIV.length);
+
+						byte[] result = new byte[commonIV.length];
+
+						int i = 0;
+						for (byte b : tmp) {
+							result[i] = (byte) (b ^ commonIV[i++]);
+						}
+
+						return result;
+					} else {
+						LOGGER.error(Error.NONCE_LENGTH_INVALID);
+						throw new IllegalArgumentException(Error.NONCE_LENGTH_INVALID);
+					}
+				} else {
+					LOGGER.error(Error.COMMON_IV_NULL);
+					throw new NullPointerException(Error.COMMON_IV_NULL);
+				}
+			} else {
+				LOGGER.error(Error.SENDER_ID_NULL);
+				throw new NullPointerException(Error.SENDER_ID_NULL);
+			}
+		} else {
+			LOGGER.error(Error.PARTIAL_IV_NULL);
+			throw new NullPointerException(Error.PARTIAL_IV_NULL);
+		}
+	}
+
+	/**
+	 * Padds the left side of the byte array paddMe with zeroes as the int
+	 * zeroes has
+	 * 
+	 * @param paddMe
+	 * @param zeroes
+	 * @return
+	 */
+	private static byte[] leftPaddingZeroes(byte[] paddMe, int zeroes) {
+		byte[] tmp = new byte[zeroes + paddMe.length];
+		Arrays.fill(tmp, 0, zeroes - 1, (byte) 0x00);
+		return tmp;
```

Comment:
Hm, I think, this creates a byte array initialized with 0s (I think, it's "zeros", so "zeroes" maybe misspelled) and the fills the array with 0s (which are already there).

So:
0. correct "zeroes" to "zeros"
1. make the method public and create unit tests.
2. check, if `System.arraycopy(paddMe, 0, tmp, zeros, paddMe.length);` does it.

Reply:
Fixed.